---
title: "iMKT using PopFly or PopHuman data"
output: 
  pdf_document:
    highlight: monochrome
    latex_engine: xelatex
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{iMKT Pipeline}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	fig.align = "center",
	fig.height = 7,
	fig.width = 7,
	collapse = TRUE,
	comment = "#>"
)
```

Brief intro about lots of data right now. Then, PopFly and PopHuman as great databases.

Therefore, iMKT package includes some functions which allow an easy retrieval and analysis of population genetics information stored in these genome browsers. Specifically, the functions permit the download of population genetics parameters computed for every gene annotation in several populations for both model species *Drosophila melanogaster* and *Homo sapiens*.

The examples in this vignette focus only on PopFly data and analysis, but the same process could be performed using human data and funcions (**loadPopHumandata()**, **PopHumanAnalysis()**, and **PopHumanData**), following the same steps described here. Recombination rate values of human data are retrieved from Bheller et al. and correspond to the sex average estimates.

&nbsp;
&nbsp;

## Loading the row data

The first step is to load the information into your working enviroment. This would allow a manual examination of the data before starting any alayses. However, this step can be skipped, and the main data object is loaded into the workspace the first time that the analysis function is called. 
To download PopFly data use the **loadPopFly()** function, without any argument. 

Keep in mind that you are downloading the complete gene information of several populations (16 *D. melanogaster* in this case and 26 in the case of *H. sapiens*), and this procces could take a while.

Once the function finishes,the **PopFlyData** object is loaded into the workspace. This object can be manually examined or used when performing the main iMKT analyses.
  
&nbsp;
&nbsp;
  
```{r popfly data, echo=TRUE, fig.width=7}
library(iMKT)
loadPopFly()
ls() ## new object created
names(PopFlyData)
```

&nbsp;
&nbsp;
  
  
Each row of the **PopFlyData** dataframe contains information regarding one gene annotation in one single population. In total, 16 populations and 13745 genes are included. Metrics for each gene contain information about segrating, divergent and analyzed sites and the Derived Allele Frequency (DAF) for neutral (4fold) and putatively selected (0fold) sites; together with some neutrality tests statistics (Standard MKT, Direction of Selection, Ka/Ks) and gene-associated recombination rate estimates (from Comeron et al.).

<!-- As we compress the Derived Allele Frequency in a column field, you will need to create the correct variables forms to pass them properly at the functions. It can be done in a few R basics lines! -->

```{r PopFly data manual retrieve, echo=FALSE, eval=FALSE}
## Preparing RAL Adh
adhRAL<-dataPopfly[dataPopfly$Name=='FBgn0000055' & dataPopfly$Pop=='RAL',]

adhRAL$DAF0f <- as.character(adhRAL$DAF0f); adhRAL$DAF4f <- as.character(adhRAL$DAF4f)
adhRAL0f<-unlist(strsplit(adhRAL$DAF0f, split=";"))
adhRAL4f <- unlist(strsplit(adhRAL$DAF4f, split=":"))
adhRAL0f <- as.numeric(adhRAL0f); adhRAL4f <- as.numeric(adhRAL0f)

f <- seq(0.05,0.95,0.1)
mi <- adhRAL$mi; m0 <- adhRAL$m0
Di <- adhRAL$di; D0 <- adhRAL$d0

dafAdhRAL <- cbind(f, adhRAL0f, adhRAL4f); dafAdhRAL <- as.data.frame(dafAdhRAL)
names(dafAdhRAL) <- c("daf","Pi","P0")
divAdhRAL <- cbind(mi, Di, m0, D0); divAdhRAL <- as.data.frame(divAdhRAL)
names(divAdhRAL) <- c("mi","Di","m0","D0")

## Preparing ZI Adh
adhZI<-dataPopfly[dataPopfly$Name=='FBgn0000055' & dataPopfly$Pop=='ZI',]

adhZI$DAF0f <- as.character(adhZI$DAF0f); adhZI$DAF4f <- as.character(adhZI$DAF4f)
adhZI0f<-unlist(strsplit(adhZI$DAF0f, split=";"))
adhZI4f <- unlist(strsplit(adhZI$DAF4f, split=":"))
adhZI0f <- as.numeric(adhZI0f); adhZI4f <- as.numeric(adhZI0f)

f <- seq(0.05,0.95,0.1)
mi <- adhZI$mi; m0 <- adhZI$m0
Di <- adhZI$di; D0 <- adhZI$d0

dafAdhZI <- cbind(f, adhZI0f, adhZI4f); dafAdhZI <- as.data.frame(dafAdhZI)
names(dafAdhZI) <- c("daf","Pi","P0")
divAdhZI <- cbind(mi, Di, m0, D0); divAdhZI <- as.data.frame(divAdhZI)
names(divAdhZI) <- c("mi","Di","m0","D0")
```
<!-- Once the data is in the proper form you can perform the test without problems -->

```{r PopFly data retrieve Adh RAL, echo=FALSE, eval=F}
standard(daf = dafAdhRAL, divergence = divAdhRAL)
DGRP(daf = dafAdhRAL, divergence = divAdhRAL,plot = TRUE)
```

```{r PopFly data retrieve Adh ZI, echo=F, eval=F}
standard(daf = dafAdhZI, divergence = divAdhZI)
DGRP(daf = dafAdhZI, divergence = divAdhZI,plot = TRUE)
```

<!-- This proccess could be tedious depending on the number of genes you want to analyze. To solve this, we just add  -->

Once the data is loaded into the workspace, analyses can be performed using the function **PopFlyAnalysis()**.

&nbsp;
&nbsp;
&nbsp;
&nbsp;

## Performing PopFly Analyses
  
The **PopFlyAnalisys()** function allows performing any MK test using a subset of PopFly data defined by custom genes and populations lists. It uses the previously loaded dataframe (PopFlyData). In addition to the **genes** and **populations** lists, the function also has the following parameters:

  - **recomb**: group genes according to recombination values (must specify number of bins). Options are: TRUE/FALSE. Recombination values (cM/Mb) from Comeron et al. 2012.
  - **bins**: number of recombination bins to compute (mandatory if recomb = TRUE)
  - **test**: which test to perform. Options include: standard (default), DGRP, FWW, asymptotic, iMK
  - **xlow**: lower limit for asymptotic alpha fit (default=0)
  - **xhigh**: higher limit for asymptotic alpha fit (default=1)
  
&nbsp;

Custom genes must be listed using FlyBase IDs (FBgn...), and the available populations from PopFly are: AM, AUS, CHB, EA, EF, EG, ENA, EQA, FR, RAL, SA, SD, SP, USI, USW, ZI.

Hence, using the parameters recomb and bins, the function allows deciding whether to analyze genes groupped by recombination bins or not.

&nbsp;
&nbsp;
  
### Example 1
  
In this first example, the analysis is focused in two genes and 2 populations, without considering gene's recombination context, and using DGRP methodology. 

The function groups polymorphism and divergence values of the custom genes, creating a new "concatenated gene" for each population of interest and performs the test defined. Then, it returns a list of lists with the default test output (DGRP in this case) for each population (RAL and ZI).

Explain output of DGRP function.

&nbsp;
&nbsp;

```{r PopFly data retrieve automated no recomb, echo=TRUE}
PopFlyAnalysis(genes=c("FBgn0000055","FBgn0003016"), pops=c("RAL","ZI"), recomb=F, test="DGRP")
```
  
&nbsp;
&nbsp;
  
### Example 2  
 
In this second example, genes from RAL population are groupped in 2 recombination bins, using recombination values from Comeron et al (ref). The test used is iMK, with xlow and xhigh values set to 0 and 0.9, respectively.

In this case, the function creates two different "concatenated" genes containing the same number of genes each (6 in this case), grouping again polymorphism and divergence values. These aggrupations are made according to the gene associated recombination rate estimates. The function returns a list of lists with the default test output (iMKT in this case) for each population (RAL) and recombination bin (1 and 2).

Explain iMK function output.

&nbsp;
&nbsp;

```{r PopFly data retrieve automated, echo=TRUE}
geneList <- c("FBgn0053196","FBgn0086906","FBgn0261836","FBgn0031617","FBgn0260965",
              "FBgn0028899","FBgn0052580","FBgn0036181","FBgn0263077","FBgn0013733",
              "FBgn0031857","FBgn0037836")

PopFlyAnalysis(genes=geneList , pops=c("RAL"), recomb=T, bins=2, test="iMK", xlow=0, xhigh=0.9)
```
  
&nbsp;
&nbsp;
