---
title: "Vignette Title"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>", fig.width = 7, fig.height = 7, fig.align = "center")
```

## The McDonald and Kreitman Test
Theoretic explanation  
Since adaptive mutations tend to be fixed quickly, they will rarely be detected as polymorphic variants but only as a divergent site (that is once fixed). Thus, the adaptive substitution rate in the genome can be inferred when there is an excess of non-synonymous divergence relative to its non-synonymous polymorphism. Accordingly, in the McDonald-Kreitman test (MKT) (McDonald 1and Kreitman, 1991) the divergence ratio (d N /d S ) is normalized by the poly- morphism ratio (π n /π s ), which allows taking into account the constraint on non- d N /d S synonymous sites and, thus, better detect adaptive substitutions (> 1). π n /π s From that one can also estimate the proportion of fixed variants that are adap- tive (α). However, the estimation of α can be biased due to the segregation of slightly deleterious non-synonymous mutations (Eyre-Walker, 2002). Given a stable population size, slightly deleterious mutations can produce an understimation of α because they tend to contribute more to polymorphism than to divergence. Because slightly deleterious substitutions tend to segregate at low frequency, its effect can be partially controlled by removing low- frequency polymorphisms from the analysis, known as Fay-Wycoff-Wu method (Fay et al., 2001). However, Charlesworth and Eyre-Walker (2008) showed that even removing low-frequency variants, the estimate of α is always down- wardly biased and only these estimates are reasonable accurate when the rate of adaptive evolution is high and the distribution of fitness effects of slightly deleterious mutations is leptokurtic (because leptokurtic distributions have a smaller proportion of polymorphisms that are slightly deleterious). The modification of the MKT introduced at the DGRP project (Mackay et al., 2012). was proposed as a better method for correcting the effect of slightly deleterious substitutions. Instead of simply removing low-frequency polymorphisms, the count of segregating sites in non-synonymous sites is separated into the num- ber of neutral variants and the number of weakly deleterious variants, allowing evaluating independently adaptive and weakly deleterious selection.

### Pipeline
The package is deposited in the official repository CRAN (ojalá), and in the BGD group github. It could be downloaded by the following methods:
```{r Loading the package}
# install.packages("iMKT") ##If CRAN  
# library(devtools) ##If github
# install_github("sergihervas/iMKT")  
  
library(iMKT)
```
The data used in this tutorial is incorporated inside the package in order to use it as tutorial, or replicate this vignettes to understand better all the package functionalities. You could access the example data and save it in your own variable, but the are loaded in your enviroment by default loading the package!
```{r Checking the example data}
mydafdata
mydivergencedata

exampleDaf<-mydafdata
exampleDiverge<-mydivergencedata
```

The package present the following functions
- standard()  
- FWW()  
- DGRP()  
- asymptoticMK()  
- iMK()  
- completeMKT()  
- loadPopFly()  
- loadPopHuman()  
- subsetPopData()  
- multipleDatasets()
- PopFlyAnalisys()  
- theme_Publication()  
- check_input()  
Each one execute a test, perform the calculation or load presets to obtain the pipeline resutls. Rembember you always can access to the help, to check more examples or the passing arguments writting ```{r}??``` and the function in your console!

#### Standard MKT
The MK test (McDonald and Kreitman, 1991) was developed to be applied to protein coding sequences, combining both divergence (D) and polymorphism (P) sites, and categorizing mutations as synonymous (P s , D S ) and non-synonymous (P n , D N ). If all mutations are either strongly deleterious or neutral, then D N /D S is expected to roughly equal P s /P n . In contrast, if positive selection is operating in the region, adaptive mutations rapidly reach fixation and thus contribute relative more to divergence than to polymorphism when compared to neutral mutations, and then D N /D S >P n /P s . Assuming that adap- tive mutations contribute little to polymorphism but substantially to divergence, the proportion of non-synonymous substitutions than have been fixed by posi- tive selection can be inferred as α = 1 - (P n /P s )(D S /D N ) (??). The significance of effect can be easily quantified using a simple 2×2 contingency table (see Table 5), using a Fischer’s exact test.
```{r Standard MKT, echo=TRUE}
standard(daf = mydafdata,divergence = mydivergencedata)
```

#### FWW correction
α estimates can be biased by the segregation of slighlty deleterious substitutions. One method to partially controlled its effects is to remove low frequency polymorphisms from the analysis, as proposed by Fay et al. (2001).
***FWW(mydafdata,mydivergencedata)*** generate the output at the console. 
```{r FWW, echo=TRUE}
FWW(daf = mydafdata,divergence = mydivergencedata)
```
You could save it in a variable and the access to different data saved inside. Check it!
```{r, echo=TRUE}
methodFWW<-FWW(daf = mydafdata,divergence = mydivergencedata)
methodFWW$Results
methodFWW$`Divergence metrics`
methodFWW$`MKT tables`

```
By default the arguments ***list_cutoffs***, pass a list of cutoffs with the following values: c(0, 0.05, 0.1). And moreover th argument  ***plot*** is setting in ***FALSE***. You can change the cutting values and switch on the plot to visulize your results!

```{r, echo=TRUE}
methodFWW<-FWW(daf = mydafdata, divergence = mydivergencedata,list_cutoff=c(0, 0.05, 0.1,0.15,0.2),plot=TRUE)
savePlotInVariable<-methodFWW$Graph
savePlotInVariable
```


#### DGRP correction
The null hypothesis of neutrality is rejected in a MKT when DN/DS > Pn/Ps , inferring adaptation, but also when P n /P s >D N /D S . In this later case, there is an excess of polymorphism relative to divergence for the non-synonymous class n, due to (i) slightly deleterious variants segregating at low frequency in the population subject to weak negative selection, which contribute to polymorphism but not to divergence, or (ii) relaxation of selection where sites previ- ously under strong or weak purifying selection have become neutral, causing an increased level of polymorphism relative to divergence Adaptive mutations and weakly deleterious selection act in opposite direc- tions on the MKT, so α will be underestimated when the two selection regime occur. Because slightly deleterious mutations tend to segregate at lower fre- quencies than do neutral mutations, they can be partially controlled for by removing low frequency polymorphisms from the analysis, generally the 5% (Fay et al., 2001). However, this method is still expected to lead to biased estimates. To take adaptive and slightly deleterious mutation mutually into account, P n , the count of segregating sites in the non-synonymous class, should be sepa- rated into the number of neutral variants and the number of weakly deleterious variants, P n = P n-neutral +P n weakly del. . If both numbers are estimated, adaptive and weakly deleterious selection can be evaluated independently. Consider the following pair of 2×2 contingency tables (Table 5): The table on the left if the standard MKT table with the theoretical counts of segregating sites and divergent sites for each cell. The table on the right contains the count of P n and P s for two-frequency categories. The estimate of the fraction of sites segregating neutrally within the MAF (minor allele frequency) < 5% (f neutral MAF<5% ) is f neutral MAF<5% =P s MAF<5% /P s . The expected number of segregating sites in the non-synonymous class which are neutral within the MAF<5% is P neutral MAF<5% =P n ×f neutral MAF<5% . The ex- pected number of neutral segregating sites in the non-synonymous class is P n neutral =P neutral MAF<5% +P n MAF>5%. To estimate α from the standard MKT table correcting by the segregationof weakly deleterious variants, we have to substitute the P n by the expected number of neutral segregating sites, P neutral . The correct estimate of α is then α = 1 - (P n neutral/Ps)(DS/DN)
```{r DGRP, echo=TRUE}
DGRP(daf = mydafdata, divergence = mydivergencedata)
```
You could save it in a variable and the access to different data saved inside. Check it!
```{r, echo=TRUE}
methodDGRP<-DGRP(daf = mydafdata, divergence = mydivergencedata)
methodDGRP$Results
methodDGRP$`Divergence metrics`
methodDGRP$`MKT tables`

```
By default the arguments ***list_cutoffs***, pass a list of cutoffs with the following values: c(0, 0.05, 0.1). And moreover th argument  ***plot*** is setting in ***FALSE***. You can change the cutting values and switch on the plot to visulize your results!

```{r, echo=TRUE}
methodDGRP<-DGRP(daf = mydafdata, divergence = mydivergencedata,list_cutoff=c(0, 0.05,0.2),plot=TRUE)
savePlotInVariable<-methodDGRP$Graph
savePlotInVariable
```

#### Asymptotic MKT
Petrov reference + explanation
```{r Asymptotic MKT, echo=TRUE}
asymptoticMK(daf = mydafdata, divergence = mydivergencedata, xlow = 0,xhigh = 0.9)
```

#### iMK
Asymptotic explanation + Sergi slightly deleterious approach
```{r iMK, echo=TRUE}
iMK(daf = mydafdata, divergence = mydivergencedata, xlow = 0, xhigh = 0.9)
```
You could save it in a variable and the access to different data saved inside. Check it!
```{r, echo=TRUE}
methodiMK<-iMK(daf = mydafdata,divergence = mydivergencedata, xlow = 0, xhigh = 0.9)
methodiMK$Results
methodiMK$`Divergence metrics`
methodiMK$`MKT tables`
```
By default the argument ***plot*** is setting in ***FALSE***. You can change the cutting values and switch on the plot to visulize your results!
```{r, echo=TRUE}
methodiMK<-iMK(daf = mydafdata, divergence = mydivergencedata, xlow = 0, xhigh = 0.9 ,plot=TRUE)
savePlotInVariable<-methodiMK$Graph
savePlotInVariable
```

If you have a bunch of data like the following, or simply have several genes datasets:
<!-- ![alt text](~/Pictures/test.png) -->
Maybe you want to perform some test or compare the test results between your datasets. You could execute the funtion multipleDatasets, putting your datasets in a directory a name them with the extensions **ID.daf.txt**/**ID.divergence.txt**. Then execute the following commands to perform the tests:
```{r multipleDatasets, eval=FALSE, include=FALSE}
# multipleDatasets(directory = directory, test = c("standard", "DGRP", "FWW","asymptotic", "iMK"), xlow = 0, xhigh = 1, fullAnalysis = TRUE/FALSE,idList = "NA")
```
The idList argument allow to the user pass a plain text file with the IDs, in the case you want to subset the analysis to just a few datasets. It is used when fullAnalysis = FALSE, list of IDs to analyze

